File: java/sdk-http/src/main/java/dev/oppex/sdk/internal/async/RateLimitedDropLogger.java
package dev.oppex.sdk.internal.async;

import java.util.concurrent.atomic.AtomicLong;
import java.util.logging.Level;
import java.util.logging.Logger;

final class RateLimitedDropLogger {
    private static final long ONE_MINUTE_MILLIS = 60000L;

    interface Clock {
        long currentTimeMillis();
    }

    private final Logger logger;
    private final Clock clock;
    private final long intervalMillis;
    private final AtomicLong intervalStartedAt;
    private final AtomicLong droppedInInterval = new AtomicLong();

    RateLimitedDropLogger(Logger logger) {
        this(logger, new Clock() {
            public long currentTimeMillis() {
                return System.currentTimeMillis();
            }
        }, ONE_MINUTE_MILLIS);
    }

    RateLimitedDropLogger(Logger logger, Clock clock, long intervalMillis) {
        this.logger = logger;
        this.clock = clock;
        this.intervalMillis = intervalMillis;
        this.intervalStartedAt = new AtomicLong(clock.currentTimeMillis());
    }

    void recordDrop() {
        droppedInInterval.incrementAndGet();
        long now = clock.currentTimeMillis();
        long started = intervalStartedAt.get();
        if (now - started < intervalMillis || !intervalStartedAt.compareAndSet(started, now)) {
            return;
        }
        long count = droppedInInterval.getAndSet(0L);
        if (count > 0L) {
            logger.log(Level.WARNING, "Dropped {0} incidents in the last minute.", Long.valueOf(count));
        }
    }
}

File: node/src/internal/async/RateLimitedDropLogger.ts
import { logger } from '../../Logger';

/** Accumulates drop counts and emits at most one summary log line per interval (60s),
 * only if drops actually occurred — avoids flooding the host application's logs with
 * one line per dropped incident during a sustained overload. */
export class RateLimitedDropLogger {
  private droppedInInterval = 0;
  private intervalStartedAt: number;

  constructor(
    private readonly intervalMs: number,
    private readonly log: (message: string) => void = (m) => logger.warn(`[oppex-sdk] ${m}`),
    private readonly now: () => number = Date.now,
  ) {
    this.intervalStartedAt = now();
  }

  recordDrop(): void {
    this.droppedInInterval++;
    const current = this.now();
    if (current - this.intervalStartedAt >= this.intervalMs) {
      const count = this.droppedInInterval;
      this.droppedInInterval = 0;
      this.intervalStartedAt = current;
      if (count > 0) {
        this.log(`Dropped ${count} incidents in the last minute.`);
      }
    }
  }
}
